src/ops/oracle.rs

```diff
@@ -1,30 +1,312 @@
-use std::sync::atomic::{AtomicU64, Ordering};
+use std::{
+    collections::HashSet,
+    sync::{Arc, Mutex},
+};
+
+use yatp::{task::callback::TaskCell, Builder, ThreadPool};
+
+use super::transaction::Transaction;
+use crate::{closer::Closer, watermark::WaterMark, AgateOptions};
+
+struct CommittedTxn {
+    ts: u64,
+    // Keeps track of the entries written at timestamp ts.
+    conflict_keys: HashSet<u64>,
+}
 
 #[derive(Default)]
+struct CommitInfo {
+    next_txn_ts: u64,
+    /// Used by managed DB.
+    discard_ts: u64,
+    /// Contains all committed writes (contains fingerprints of keys written
+    /// and their latest commit counter).
+    committed_txns: Vec<CommittedTxn>,
+    last_cleanup_ts: u64,
+}
+
+impl CommitInfo {
+    fn cleanup_committed_transactions(&mut self, is_managed: bool, read_mark: Arc<WaterMark>) {
+        let max_read_ts = {
+            if is_managed {
+                self.discard_ts
+            } else {
+                read_mark.done_until()
+            }
+        };
+
+        assert!(max_read_ts >= self.last_cleanup_ts);
+
+        if max_read_ts == self.last_cleanup_ts {
+            return;
+        }
+        self.last_cleanup_ts = max_read_ts;
+
+        self.committed_txns.retain(|txn| txn.ts > max_read_ts);
+    }
+
+    fn has_conflict(&self, txn: &Transaction) -> bool {
+        let reads = txn.reads.lock().unwrap();
+
+        if reads.is_empty() {
+            false
+        } else {
+            // If the committed_txn.ts is less than txn.read_ts that implies that the
+            // committed_txn finished before the current transaction started.
+            // We don't need to check for conflict in that case.
+            // This change assumes linearizability. Lack of linearizability could
+            // cause the read ts of a new txn to be lower than the commit ts of
+            // a txn before it.
+            self.committed_txns
+                .iter()
+                .filter(|committed_txn| committed_txn.ts > txn.read_ts)
+                .any(|committed_txn| {
+                    reads
+                        .iter()
+                        .any(|read| committed_txn.conflict_keys.contains(read))
+                })
+        }
+    }
+}
+
 pub struct Oracle {
```

Comment:
What's the oracle used for?

Reply:
It is used for transaction to get timestamp when reading or committing.

Reply:
add a comment